transform_constantfold.ml

```diff
@@ -1,111 +1,130 @@
+    | Call (x, e, args) ->
+      let args' = List.map (fun arg ->
+        match arg with
+        | Arg_by_val e -> Arg_by_val (prop_fold e)
+        | Arg_by_ref _ -> arg) args
+      in
+      (penv, Call (x, prop_fold e, args'))
+    | Return e -> (penv, Return (prop_fold e))
+    | Assign (x, e) -> (penv, Assign (x, prop_fold e))
+    | Branch (e, l1, l2) -> (penv, Branch (prop_fold e, l1, l2))
+    | Print e -> (penv, Print (prop_fold e))
+    | Osr (e, tf, tv, tl, osr_env) ->
+      let osr_env' = List.map (fun osr_def ->
         match osr_def with
-        | Osr_const (y, e) -> Osr_const (y, replace e)
-        | Osr_mut (y, e) -> Osr_mut (y, replace e)
-        | Osr_mut_ref (y, z) -> if x=z then Osr_mut (y, Simple l) else osr_def
-        | Osr_mut_undef y -> Osr_mut_undef y) env
+        | Osr_const (y, e) -> Osr_const (y, prop_fold e)
+        | Osr_mut (y, e) -> Osr_mut (y, prop_fold e)
+        | Osr_mut_ref _ | Osr_mut_undef _ -> osr_def) osr_env
       in
-      Osr (replace exp, tf, tv, tl, env')
-    | Drop y
-    | Decl_mut (y, None)
-    | Clear y
-    | Read y ->
-      assert (x <> y);
-      instr
-    | Label _ | Goto _ | Comment _ -> instr in
+      (penv, Osr (prop_fold e, tf, tv, tl, osr_env'))
+    | Stop e -> (penv, Stop (prop_fold e))
+    | Drop x ->
+      (* `x` no longer in scope, so remove it from the environment. *)
+      (Prop_env.remove x penv, instr)
+    | Decl_mut (_, None) | Clear _ | Read _
+    | Label _ | Goto _ | Comment _ -> (penv, instr)
+  in
 
-  (* Finds the target pcs to perform constant propagation. *)
-  let rec find_targets instrs x worklist acc =
+  (* Perform the optimization over the functions.
+   * `instrs` is the instruction stream to optimize, `worklist` is the worklist,
+   * and `seen` is a PcSet that keeps track of PCs we already processed (to
+   * avoid infinite looping).
+   *
+   * `normalize` returns a pair of the new environment and new instruction.
+   * That new environment must be used for all successors of the current
+   * instruction, which is why it is included in the worklist.
+   * *)
+  let rec work instrs worklist seen =
     let open Analysis in
     match worklist with
-    | [] -> acc
-    | pc :: rest ->
-      begin match[@warning "-4"] instrs.(pc) with
-      | Drop y when x = y ->
-        (* `x` is now out of scope, but continue searching the `rest` of the
-           worklist. *)
-        find_targets instrs x rest acc
-      | instr ->
-        if PcSet.mem pc acc then
-          (* Already seen this pc, but continue searching the `rest` of the
-             worklist. *)
-          find_targets instrs x rest acc
-        else
-          (* Add the current `pc` to the accumulator and update the worklist
-             with our successors. *)
-          let succs = successors_at instrs pc in
-          find_targets instrs x (succs @ rest) (PcSet.add pc acc)
-      end
+    | [] -> Some instrs
```

Comment:
I was hoping to avoid tracking the changed status (either threading the variable everywhere, or using a mutable reference). In fact, this proposal was designed to avoid that, but this was before the new API.

But I guess it's unavoidable now, so I'll fix it.